### ISSUE TITLE
There is a heap-use-after-free in systemd

### ISSUE THREAD

**Issue by evverx** [systemd/systemd#23312]
**systemd version the issue has been seen with**
97d670a4297d4d3

It was found with `dfuzzer`
```sh
==1==ERROR: AddressSanitizer: heap-use-after-free on address 0x60b000376410 at pc 0x7f03a8ca1fec bp 0x7ffc0faa3c00 sp 0x7ffc0faa3bf8
READ of size 8 at 0x60b000376410 thread T0 (systemd)
    #0 0x7f03a8ca1feb in sd_bus_track_count ../src/libsystemd/sd-bus/bus-track.c:291
    #1 0x7f03aa658089 in unit_may_gc ../src/core/unit.c:419
    #2 0x7f03aa6596a9 in unit_add_to_gc_queue ../src/core/unit.c:489
    #3 0x7f03aa4002b6 in job_uninstall ../src/core/job.c:154
    #4 0x7f03aa65de3f in unit_free ../src/core/unit.c:704
    #5 0x7f03aa495b07 in manager_clear_jobs_and_units ../src/core/manager.c:1455
    #6 0x7f03aa4ae6ac in manager_reload ../src/core/manager.c:3366
    #7 0x41c4b6 in invoke_main_loop ../src/core/main.c:1927
    #8 0x42477d in main ../src/core/main.c:2957
    #9 0x7f03a718ff1f in __libc_start_call_main (/lib64/libc.so.6+0x40f1f)
    #10 0x7f03a718ffcf in __libc_start_main_alias_1 (/lib64/libc.so.6+0x40fcf)
    #11 0x4072e4 in _start (/usr/lib/systemd/systemd+0x4072e4)

0x60b000376410 is located 32 bytes inside of 104-byte region [0x60b0003763f0,0x60b000376458)
freed by thread T0 (systemd) here:
    #0 0x7f03aafdb627 in free (/lib64/libasan.so.6+0xae627)
    #1 0x7f03a8ca00bf in track_free ../src/libsystemd/sd-bus/bus-track.c:161
    #2 0x7f03a8ca0482 in sd_bus_track_unref ../src/libsystemd/sd-bus/bus-track.c:164
    #3 0x7f03aa65d4fc in unit_free ../src/core/unit.c:688
    #4 0x7f03aa495b07 in manager_clear_jobs_and_units ../src/core/manager.c:1455
    #5 0x7f03aa4ae6ac in manager_reload ../src/core/manager.c:3366
    #6 0x41c4b6 in invoke_main_loop ../src/core/main.c:1927
    #7 0x42477d in main ../src/core/main.c:2957
    #8 0x7f03a718ff1f in __libc_start_call_main (/lib64/libc.so.6+0x40f1f)

previously allocated by thread T0 (systemd) here:
    #0 0x7f03aafdbad7 in calloc (/lib64/libasan.so.6+0xaead7)
    #1 0x7f03a8c9eefc in sd_bus_track_new ../src/libsystemd/sd-bus/bus-track.c:130
    #2 0x7f03aa341d3f in bus_unit_allocate_bus_track ../src/core/dbus-unit.c:2542
    #3 0x7f03aa342068 in bus_unit_track_add_sender ../src/core/dbus-unit.c:2572
    #4 0x7f03aa3318d5 in bus_unit_method_ref ../src/core/dbus-unit.c:659
    #5 0x7f03a8c60c86 in method_callbacks_run ../src/libsystemd/sd-bus/bus-objects.c:406
    #6 0x7f03a8c6a486 in object_find_and_run ../src/libsystemd/sd-bus/bus-objects.c:1310
    #7 0x7f03a8c6bdd0 in bus_process_object ../src/libsystemd/sd-bus/bus-objects.c:1440
    #8 0x7f03a8ccaf6f in process_message ../src/libsystemd/sd-bus/sd-bus.c:2962
    #9 0x7f03a8ccb486 in process_running ../src/libsystemd/sd-bus/sd-bus.c:3004
    #10 0x7f03a8cce2e7 in bus_process_internal ../src/libsystemd/sd-bus/sd-bus.c:3224
    #11 0x7f03a8cce481 in sd_bus_process ../src/libsystemd/sd-bus/sd-bus.c:3251
    #12 0x7f03a8cd1b82 in io_callback ../src/libsystemd/sd-bus/sd-bus.c:3602
    #13 0x7f03a8e89bf8 in source_dispatch ../src/libsystemd/sd-event/sd-event.c:3591
    #14 0x7f03a8e9240f in sd_event_dispatch ../src/libsystemd/sd-event/sd-event.c:4175
    #15 0x7f03a8e93573 in sd_event_run ../src/libsystemd/sd-event/sd-event.c:4236
    #16 0x7f03aa4a9bf6 in manager_loop ../src/core/manager.c:3068
    #17 0x41c11c in invoke_main_loop ../src/core/main.c:1894
    #18 0x42477d in main ../src/core/main.c:2957
    #19 0x7f03a718ff1f in __libc_start_call_main (/lib64/libc.so.6+0x40f1f)

SUMMARY: AddressSanitizer: heap-use-after-free ../src/libsystemd/sd-bus/bus-track.c:291 in sd_bus_track_count
Shadow bytes around the buggy address:
  0x0c1680066c30: fd fd fd fd fa fa fa fa fa fa fa fa fd fd fd fd
  0x0c1680066c40: fd fd fd fd fd fd fd fd fd fa fa fa fa fa fa fa
  0x0c1680066c50: fa fa fd fd fd fd fd fd fd fd fd fd fd fd fd fa
  0x0c1680066c60: fa fa fa fa fa fa fa fa fd fd fd fd fd fd fd fd
  0x0c1680066c70: fd fd fd fd fd fd fa fa fa fa fa fa fa fa fd fd
=>0x0c1680066c80: fd fd[fd]fd fd fd fd fd fd fd fd fa fa fa fa fa
  0x0c1680066c90: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c1680066ca0: fd fd fa fa fa fa fa fa fa fa fd fd fd fd fd fd
  0x0c1680066cb0: fd fd fd fd fd fd fd fa fa fa fa fa fa fa fa fa
  0x0c1680066cc0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fa fa
  0x0c1680066cd0: fa fa fa fa fa fa fd fd fd fd fd fd fd fd fd fd
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==1==ABORTING
```

cc @mrc0mmand

**Comment by evverx** [comment]
@mrc0mmand I added "fail", "flush", "ignore-dependencies", "ignore-requirements", "isolate", "replace", "replace-irreversibly" to the "seed" corpus and blocked a lot of interfaces. I think it would be great if it was possible to pass a file with meaningful strings instead of rebuilding and reinstalling `dfuzzer` over and over again :-)

**Comment by mrc0mmand** [comment]
> @mrc0mmand I added "fail", "flush", "ignore-dependencies", "ignore-requirements", "isolate", "replace", "replace-irreversibly" to the "seed" corpus and blocked a lot of interfaces. I think it would be great if it was possible to pass a file with meaningful strings instead of rebuilding and reinstalling `dfuzzer` over and over again :-)

Given how useful this has proven to be, I'll stash my property-related stuff for now and try to focus on this instead :-)

**Comment by yuwata** [comment]
Fix is waiting in #23313.
